modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/dht/GridCachePartitionsUpdateCountersAndSizeTest.java

```diff
@@ -0,0 +1,243 @@
+        testLog.registerListener(lsnrCnt);
+        testLog.registerListener(lsnrSizeCnt);
+
+        // Populate cache to increment update counters.
+        for (int i = 0; i < 1000; i++)
+            ignite.cache(CACHE_NAME).put(i, i);
+
+        if (cnt)
+            // Modify update counter for some partition.
+            ignite.cachex(CACHE_NAME).context().topology().localPartitions().get(0).updateCounter(100500L);
+
+        if (size)
+            // Modify update size for some partition.
+            ignite.cachex(CACHE_NAME).context().topology().localPartitions().get(0).dataStore()
+                .clear(ignite.cachex(CACHE_NAME).context().cacheId());
+
+        // Trigger exchange.
+        startGrid(3);
+
+        awaitPartitionMapExchange();
+
+        // Nothing should happen (just log error message) and we're still able to put data to corrupted cache.
+        ignite.cache(CACHE_NAME).put(0, 0);
+
+        if (cnt && !size)
+            assertTrue("Counters inconsistent message not found", lsnrCnt.check());
+
+        if (!cnt && size)
+            assertTrue("Size inconsistent message not found", lsnrSize.check());
+
+        if (cnt && size)
+            assertTrue("Both counters and sizes message not found", lsnrSizeCnt.check());
+
+        if (!cnt && !size)
+            assertFalse("Counters and Size inconsistent message found!", lsnrSize.check() && lsnrCnt.check()
+                && lsnrSizeCnt.check());
+    }
+
+    /**
+     *  1. Only partition counters are inconsistent.
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testValidationfPartitionCountersInconsistent() throws Exception {
+        startThreeNodesGrid(true, false);
+    }
+
+    /**
+     *  2. Only partition size are inconsistent.
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testValidationfPartitionSizeInconsistent() throws Exception {
+        startThreeNodesGrid(false, true);
+    }
+
+    /**
+     *  3. Partition counters and size are inconsistent.
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testValidationBothPartitionSizesAndCountersAreInconsistent() throws Exception {
+        startThreeNodesGrid(true, true);
+    }
+
+    /**
+     *  4. No one partition counters and size are inconsistent.
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testValidationBothPatririonSixeAndCountersAreConsistent() throws Exception {
+        startThreeNodesGrid(false, false);
+    }
+
+    /**
+     *  Overraided class LogListener.
+     * @throws Exception If failed.
+     */
+    private static class SizeCounterLogListener extends LogListener {
+
+        @Override public boolean check() {
+            return false;
+        }
+
+        @Override public void reset() {
+
+        }
+
+        // Search specific string pattern and add value of counters and sizes to arrays
+        @Override public void accept(String s) {
```

Comment:
Empty javadoc.

Reply:
It's a bit confusing and not obvious to me. 
It can simplify the implementation, for example, passing regexp to get the necessary fields, and then checking all the necessary numbers in the body of the test method.

Reply:
it work

Reply:
Can I use regexp and get groups that I can check for?
https://www.tutorialspoint.com/javaregex/javaregex_capturing_groups.htm